### PR TITLE
Add possibility to easier override HP and breath engine logic by Lua

### PR DESCRIPTION
**Pull request by sfence**
- Add the possibility to disable C++-based HP and breath logic.
- Add per player mask which can be used to enable/disable 
- This fixes issue #8042.

## To do

This PR is a Ready for Review

- [x] Get feedback
- [x] Apply feedback
- [x] Add documentation

## How to test

Run devtest game and change the `engine_mask` value via the included item for properties change.

### CODE REVIEW

**Comment by lhofhansl** [comment]
Wasn't there some discussion to move the entire logic to Lua?

**Comment by sfence** [comment]
> Wasn't there some discussion to move the entire logic to Lua?

Yes, in #14125.
But this solution should target to allow full backward compatibility and moving logic to Lua as well.
I also think that the C++ solution has better performance than the Lua alternative. So, why not keep it where it is possible?

**Comment by Zughy** [comment]
>  So, why not keep it where it is possible?

Because, contrary to HPs, it's not something the majority of games need. It should be eradicated from the engine imho, it's only a burden for core devs

**Comment by sfence** [comment]
> Because, contrary to HPs, it's not something the majority of games need. It should be eradicated from the engine imho, it's only a burden for core devs

I understand.

But, before some time I created some mod that was doing something like breath in Lua. The problem with it was, that it did not work smoothly because of the non-constant time difference between steps.

So, I think that the use of C++ HP and breath engine whatever is possible will look better from the player's side as a result.

File: src/object_properties.cpp
Comment (appgurueu) [inline src/object_properties.cpp:252]: `f32` (for consistency)

File: src/object_properties.h
Comment (appgurueu) [inline src/object_properties.h:78]: 4 new properties for every object for this seems somewhat wasteful. Shouldn't this apply just to players? Modders can implement their own logic for luaentities anyways.

File: src/script/cpp_api/s_player.cpp
Comment (appgurueu) [inline src/script/cpp_api/s_player.cpp:273]: What's with this outcommented code?

File: src/server/player_sao.cpp
Comment (appgurueu) [inline src/server/player_sao.cpp:445]: Delete?

**Review by appgurueu — COMMENTED** [review]
I think this PR is overkill for de-hardcoding breath; all you need for that would be a simple flag that tells the engine to stop managing breath so that the modder can take over. (In fact, this PR goes a bit in the opposite direction of #14125 even IMO: It adds breath to entities.)

I would suggest splitting this up into a PR which lets you disable engine breath handling for players, which will basically fix #8042 and half of #14125 and should be rather uncontroversial, and another PR for all the features, esp. adding node damage to entities etc. (which I'm frankly not sure belongs into the engine).

> So, I think that the use of C++ HP and breath engine whatever is possible will look better from the player's side as a result.

I'm not so sure about this. If I see this correctly, this PR is entirely serverside and uses no server features not available to Lua, so why would it have clientside prediction advantages over a Lua-only solution, which would be entirely serverside as well?
(By the way, I think breath was moved to the server to combat cheating a rather long while ago?)

**Comment by sfence** [comment]
So, it looks like the idea of using a mask to add the ability to partially disable SAO damage/breath logic in combination with Lua callback does not sound interesting for Minetest developers.

So, this idea is probably a step aside.

So, the best way is probably to abandon this idea of a solution. Am I right?

**Comment by lhofhansl** [comment]
> So, the best way is probably to abandon this idea of a solution. Am I right?

Not the idea. I really like @appgurueu suggestion above. Add a flag to disable engine's breath management, and separate for the Lua API.
I think for now we can leave the C++ logic in place, and perhaps remove it later.

Please do not be discouraged, we like contributions that make modding easier/possible.

File: src/script/cpp_api/s_player.cpp
Comment (grorp) [inline src/script/cpp_api/s_player.cpp:264]: Adding an `on_playerstep` callback is unrelated to allowing to disable engine features. Please remove the change from this PR. You can put it into a separate PR if you want.

**Review by grorp — CHANGES_REQUESTED** [review]
This PR still contains the change to add node damage to entities. That doesn't belong in this PR and breaks backwards compatibility if enabled by default. Please remove it.

**Comment by sfence** [comment]
> This PR still contains the change to add node damage to entities. That doesn't belong in this PR and breaks backwards compatibility if enabled by default. Please remove it.

It is related to bug #9372. But yes, it breaks back compatibility. Will it be satisfying to disable it by default?
So backward compatibility will not be broken and #9372 will be also fixed with this PR.

**Comment by grorp** [comment]
Well, actually I'd be fine with adding node damage to entities as long as it's disabled by default. It would save modders from having to write their own node damage reimplementations in case they just want entities to behave the same as players. However, this seems to be a controversial idea considering the comments in #9372. Maybe it would be better to keep this PR simple and uncontroversial by just including the "allow disabling hardcoded engine features" change.

File: doc/lua_api.md
Comment (grorp) [inline doc/lua_api.md:8537]: Bitmasks seem like a foreign thing in a simple language like Lua. I think we don't have a bitmask anywhere else in our Lua API.

Comment (appgurueu) [reply doc/lua_api.md:8537]: We have bitsets in a couple places, the `style` property of HUD `text` elements for example ("Bitfield with 1 = bold, 2 = italic, 4 = monospace"), but I agree that this should ideally be hidden from the modder: A Lua-friendly API should simply provide boolean fields unless there is a very strong reason not to (like performance when considering `param2` in nodes for example), which I don't see here - object properties could still use a bitset under the hood. That said

> Maybe it would be better to keep this PR simple and uncontroversial by just including the "allow disabling hardcoded engine features" change.

Yes, I would prefer this. I think it's the best idea to leave the (re)implementation of these features for entities up to modders.

**Comment by Zughy** [comment]
@sfence reminder

File: src/script/common/c_content.cpp
Comment (grorp) [inline src/script/common/c_content.cpp:325]: `x ? true : false` is redundant?

File: src/script/common/c_content.cpp
Comment (grorp) [inline src/script/common/c_content.cpp:297]: Inserting `read_engine_mask` and `push_engine_mask` in-between `object_property_keys` and `read_object_properties` isn't very sensible. I'd rather add the new functions after `push_object_properties`.

File: src/script/common/c_content.cpp
Comment (grorp) [inline src/script/common/c_content.cpp:606]: >  Remember to update object_property_keys above when adding a new property

File: src/script/common/c_content.cpp
Comment (grorp) [inline src/script/common/c_content.cpp:605]: inconsistent spacing (you effectively removed the empty line before the comment)

**Review by grorp — COMMENTED** [review]
The new table-based form for `engine_mask` isn't documented. Also, is there a reason for exposing the underlying bitmask to the Lua API by allowing to specify `engine_mask` as a number?

I again suggest the https://dev.minetest.net/Code_style_guidelines, for example this:

> When breaking conditionals, indent following lines of the conditional with two tabs and the statement body with one tab.

File: src/script/common/c_content.cpp
Comment (grorp) [inline src/script/common/c_content.cpp:582]: why does this even return a bool?

File: doc/lua_api.md
Comment (grorp) [inline doc/lua_api.md:8855]: just spelling and stuff

```suggestion
    engine_mask = {
    -- This can be used to enable/disable some parts of the engine.
      drowning = true,
      -- Drowning (only available for players)
      breathing = true,
      -- Breathing (only available for players)
      node_hurt = true,
      -- Node damage (only available for players)
    },
```

File: src/script/common/c_content.cpp
Comment (grorp) [inline src/script/common/c_content.cpp:51]: looks like my suggestion where to put `push_engine_mask` and `read_engine_mask` wasn't very good either 🙃

**Review by grorp — COMMENTED** [review]
I'm not sure about the naming, but I currently have no better naming proposal.

Tested. The default value behavior didn't match my expectations. When not specifying `engine_mask`, all flags default to true. When specifying `engine_mask` as `{breathing = false}`, the other flags default to false, not to true as I would expect. Defaulting to true in this case would also result in better forward-compatibility if more flags are added in the future.

File: src/script/common/c_content.cpp
Comment (sfence) [reply src/script/common/c_content.cpp:51]: I can move it back if you think it looks better before moving.

File: src/script/common/c_content.cpp
Comment (sfence) [reply src/script/common/c_content.cpp:582]: copy paste resudium

File: src/script/common/c_content.cpp
Comment (grorp) [reply src/script/common/c_content.cpp]: doesn't matter

**Comment by grorp** [comment]
> Tested. The default value behavior didn't match my expectations. When not specifying `engine_mask`, all flags default to true. When specifying `engine_mask` as `{breathing = false}`, the other flags default to false, not to true as I would expect. Defaulting to true in this case would also result in better forward-compatibility if more flags are added in the future.

Since you replied to everything else, I'm repeating this in case you missed it.

**Comment by sfence** [comment]
> Tested. The default value behavior didn't match my expectations. When not specifying `engine_mask`, all flags default to true. When specifying `engine_mask` as `{breathing = false}`, the other flags default to false, not to true as I would expect. Defaulting to true in this case would also result in better forward-compatibility if more flags are added in the future.

In Lua `nil` and `false`  are commonly expected to have similar/same effect. So it can be confusing to replace `nil` with `true`.
Maybe I can change the logic of the `engine_mask` table not to enable part of engine logic, but to disable part of engine logic.
```
engine_logic = {
  no_breathing = true,
   ...
}
```
Does it sound better?

**Comment by sfence** [comment]
Ok, `nil` is used with no change meaning in other parts of the `Lua` API tool. So it should be understandable. The previous version of `read_engine_mask` was also potentially buggy, so it should be better now.

**Review by grorp — APPROVED** [review]
Works as expected now.

I think defaulting to the old values makes sense because that's how top-level object properties work too.

**Comment by SmallJoker** [comment]
I like this PR and was about to submit my review when I got an idea. Wouldn't it be more helpful to control these damage ticks with floats for each? That would allow arbitrary delays, or also disabling it with values < 0.

**Comment by sfence** [comment]
> I like this PR and was about to submit my review when I got an idea. Wouldn't it be more helpful to control these damage ticks with floats for each? That would allow arbitrary delays, or also disabling it with values < 0.

This possibility was included but rejected. So I removed it.

After this PR, I want to create PR for the player `on_step` callback. It will add the possibility to implement custom delays, breath/damage, and similar logic in the player `on_step` Lua callback.

**Review by appgurueu — COMMENTED** [review]
Tested, works:

- [x] enabling/disabling drowning works
- [x] enabling/disabling breathing works
- [x] enabling/disabling node damage works

I believe it would be cleaner for this to be a player-SAO field rather than an object property. I've sent you a pull request to refactor this appropriately (https://github.com/sfence/minetest/pull/3). I'd also like to hear what @grorp thinks.

Besides this, I only have minor comments.

(There's an edge case I consider slightly odd, but which I don't think needs to be addressed (in particular because modders can just turn off the system entirely if they are unhappy with it): When you disable breathing and enable drowning, you only lose health *while you are actively drowning*. When you "stop drowning" - but stay at zero air - you do not lose health. Intuitively, this doesn't really make sense to me: When you have no air, shouldn't you continue to lose health?)

**Comment by sfence** [comment]
@appgurueu 
> I believe it would be cleaner for this to be a player-SAO field rather than an object property. I've sent you a pull request to refactor this appropriately ([sfence#3](https://github.com/sfence/minetest/pull/3)). I'd also like to hear what @grorp thinks.

I am planning to use the same field for fix #9372. In that case, is it better to have two separate SAO fields in player-SAO and entity-SAO?

File: src/script/common/c_content.cpp
Comment (sfence) [reply src/script/common/c_content.cpp]: outdated

File: src/script/common/c_content.cpp
Comment (sfence) [reply src/script/common/c_content.cpp]: outdated

File: src/script/common/c_content.cpp
Comment (sfence) [reply src/script/common/c_content.cpp]: outdated

File: src/object_properties.h
Comment (sfence) [reply src/object_properties.h]: outdated

**Comment by sfence** [comment]
@grorp I take @appgurueu ideas into this PR. So please recheck your approval.

**Comment by appgurueu** [comment]
I'm happy with this now; it has my approval. grorp can mark this as "two approvals" if he's happy too. Since we're in feature freeze, I've put this on the 5.10 milestone.

**Comment by grorp** [comment]
Looks alright.

**Comment by sfence** [comment]
Rebased -> easier to build now -> no manual irrlicht cloning.

**Comment by SmallJoker** [comment]
I don't think there was much to fix up in the rebase but I think it's better safe than sorry, thus re-requested a review.
Thank you.

**Review by appgurueu — APPROVED** [review]
Tested the three flags, everything still works as expected. Code also still makes sense.

**Comment by nerzhul** [comment]
> > So, why not keep it where it is possible?
> 
> Because, contrary to HPs, it's not something the majority of games need. It should be eradicated from the engine imho, it's only a burden for core devs

generally i'm for C++ way, but on this specific case i agree this logic has nothing to do on the engine, but it's pretty hard, the HP logic is everywhere.
For the breath logic it should be easier to port

**Pull request merged**
